[Python][Flaky Test] test_sync_resubscribe_after_connection_kill - TimeoutError in cluster mode

### Test Name

test_sync_resubscribe_after_connection_kill_exact_channels
test_sync_resubscribe_after_connection_kill_patterns

### Test Location

tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels
tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_patterns

### Failure Permlink

https://github.com/valkey-io/valkey-glide/actions/runs/28420121702/job/84211287982#logs

### Frequency

Observed in scheduled CI run on 2026-06-30. Failed in all PubSub test jobs (cluster mode True only). All subscription methods (Config, Lazy, Blocking) and all message read methods (Async, Sync, Callback) are affected.

### Steps to Reproduce

1. Run the FMT Python workflow on schedule
2. Observe failure in Python PubSubTests jobs with cluster mode enabled

### System Information

GitHub Actions (ubuntu-24.04, x86_64 and aarch64)

### Language and Version

Python 3.9 - 3.14

### Engine Version

All engine versions (6.2, 7.0, 7.2, 8.0, 8.1, 9.0)

### Logs

```
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Config-MessageReadMethod.Async-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Config-MessageReadMethod.Sync-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Config-MessageReadMethod.Callback-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Lazy-MessageReadMethod.Async-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Lazy-MessageReadMethod.Sync-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Lazy-MessageReadMethod.Callback-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Blocking-MessageReadMethod.Async-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Blocking-MessageReadMethod.Sync-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_exact_channels[SubscriptionMethod.Blocking-MessageReadMethod.Callback-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_patterns[SubscriptionMethod.Config-MessageReadMethod.Async-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_patterns[SubscriptionMethod.Config-MessageReadMethod.Sync-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_patterns[SubscriptionMethod.Lazy-MessageReadMethod.Async-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_patterns[SubscriptionMethod.Lazy-MessageReadMethod.Sync-True] - glide_shared.exceptions.TimeoutError: timed out
FAILED tests/sync_tests/test_sync_pubsub.py::TestSyncPubSub::test_sync_resubscribe_after_connection_kill_patterns[SubscriptionMethod.Lazy-MessageReadMethod.Callback-True] - glide_shared.exceptions.TimeoutError: timed out

Pattern: Only cluster mode (True) fails. Standalone mode (False) passes for the same test variants.
```

### Screenshots

_No response_

### Glide Version

main

### Expected Behavior

Test should pass reliably. After a connection kill, the client should automatically resubscribe to channels/patterns in cluster mode.

### Actual Behavior

The test times out when attempting to verify resubscription after connection kill in cluster mode. All subscription methods (Config, Lazy, Blocking) and message read methods (Async, Sync, Callback) are affected, but only in cluster mode. Standalone mode passes consistently. This indicates that the cluster-mode resubscription logic after connection recovery may be stuck or the messages are not being delivered after resubscription.

### Possible Fixes

Since only cluster mode fails:
1. The cluster reconnection logic may not properly trigger resubscription for killed connections
2. The slot refresh after connection kill may interfere with pubsub resubscription
3. Timeout values for cluster reconnection + resubscription may need to be increased
4. Related to #6361 - may share same root cause with other PubSub cluster timeout issues

--- a/python/tests/sync_tests/test_sync_pubsub.py
+++ b/python/tests/sync_tests/test_sync_pubsub.py
@@ -3588,20 +3588,45 @@ class TestSyncPubSub:
             # Kill connections - this should trigger reconnection
             kill_connections(publishing_client, None)
 
-            # Give some time for connection to reconnect
-            time.sleep(2)
-
-            # Wait for subscriptions to be re-established (need to poll since reconnection is async)
+            # Wait for subscriptions to be re-established after reconnection.
+            # In cluster mode, reconnection involves multiple nodes, topology
+            # refresh, and resubscription which can take significantly longer.
+            resubscribe_timeout = 15.0 if cluster_mode else 5.0
             sync_wait_for_subscription_state(
-                listening_client, expected_channels={channel}, timeout_sec=5.0
+                listening_client,
+                expected_channels={channel},
+                timeout_sec=resubscribe_timeout,
             )
 
-            # Verify subscription still works after reconnection
-            publishing_client.publish(message_after, channel)
-            time.sleep(1)
+            # After resubscription state is confirmed, the server-side
+            # subscription may still need a brief moment to become fully active
+            # (especially in cluster mode). Use a publish-and-retry loop to
+            # handle this race condition reliably.
+            msg_after = None
+            max_publish_attempts = 5
+            for attempt in range(max_publish_attempts):
+                publishing_client.publish(message_after, channel)
 
-            msg_after = sync_get_message_by_method(
-                method, listening_client, callback_messages, 1
+                # Poll for the message using try_get or callback check
+                poll_deadline = time.time() + 3.0
+                while time.time() < poll_deadline:
+                    if method == MethodTesting.Callback:
+                        if len(callback_messages) >= 2:
+                            msg_after = decode_pubsub_msg(callback_messages[1])
+                            break
+                    else:
+                        result = listening_client.try_get_pubsub_message()
+                        if result is not None:
+                            msg_after = decode_pubsub_msg(result)
+                            break
+                    time.sleep(0.1)
+
+                if msg_after is not None:
+                    break
+
+            assert msg_after is not None, (
+                f"Failed to receive message after reconnection "
+                f"({max_publish_attempts} publish attempts)"
             )
             assert msg_after.message == message_after
             assert msg_after.channel == channel
@@ -3670,20 +3695,45 @@ class TestSyncPubSub:
             # Kill connections - this should trigger reconnection
             kill_connections(publishing_client, None)
 
-            # Give some time for connection to reconnect
-            time.sleep(2)
-
-            # Wait for subscriptions to be re-established (need to poll since reconnection is async)
+            # Wait for subscriptions to be re-established after reconnection.
+            # In cluster mode, reconnection involves multiple nodes, topology
+            # refresh, and resubscription which can take significantly longer.
+            resubscribe_timeout = 15.0 if cluster_mode else 5.0
             sync_wait_for_subscription_state(
-                listening_client, expected_patterns={pattern}, timeout_sec=5.0
+                listening_client,
+                expected_patterns={pattern},
+                timeout_sec=resubscribe_timeout,
             )
 
-            # Verify subscription still works after reconnection
-            publishing_client.publish(message_after, channel)
-            time.sleep(1)
+            # After resubscription state is confirmed, the server-side
+            # subscription may still need a brief moment to become fully active
+            # (especially in cluster mode). Use a publish-and-retry loop to
+            # handle this race condition reliably.
+            msg_after = None
+            max_publish_attempts = 5
+            for attempt in range(max_publish_attempts):
+                publishing_client.publish(message_after, channel)
 
-            msg_after = sync_get_message_by_method(
-                method, listening_client, callback_messages, 1
+                # Poll for the message using try_get or callback check
+                poll_deadline = time.time() + 3.0
+                while time.time() < poll_deadline:
+                    if method == MethodTesting.Callback:
+                        if len(callback_messages) >= 2:
+                            msg_after = decode_pubsub_msg(callback_messages[1])
+                            break
+                    else:
+                        result = listening_client.try_get_pubsub_message()
+                        if result is not None:
+                            msg_after = decode_pubsub_msg(result)
+                            break
+                    time.sleep(0.1)
+
+                if msg_after is not None:
+                    break
+
+            assert msg_after is not None, (
+                f"Failed to receive message after reconnection "
+                f"({max_publish_attempts} publish attempts)"
             )
             assert msg_after.message == message_after
             assert msg_after.channel == channel

--- a/python/tests/sync_tests/test_sync_pubsub.py
+++ b/python/tests/sync_tests/test_sync_pubsub.py
@@ -47,6 +47,60 @@ from tests.utils.utils import (
     sync_check_if_server_version_lt,
 )
 
+# Defaults for the publish-and-retry loop used to verify message delivery after
+# a resubscription. After subscription state is confirmed as restored, the
+# server-side subscription may still need a brief moment to become fully active
+# (especially in cluster mode), so a published message can be silently dropped.
+_MAX_PUBLISH_ATTEMPTS = 5
+_PUBLISH_POLL_TIMEOUT_SEC = 3.0
+_PUBLISH_POLL_INTERVAL_SEC = 0.1
+
+
+def _publish_and_wait_for_message(
+    publishing_client,
+    message: str,
+    channel: str,
+    listening_client,
+    method: MethodTesting,
+    callback_messages: List[PubSubMsg],
+    expected_idx: int,
+    max_attempts: int = _MAX_PUBLISH_ATTEMPTS,
+    poll_timeout: float = _PUBLISH_POLL_TIMEOUT_SEC,
+    poll_interval: float = _PUBLISH_POLL_INTERVAL_SEC,
+) -> Optional[PubSubMsg]:
+    """
+    Publish ``message`` to ``channel`` and poll for its receipt, re-publishing if
+    it is not received within ``poll_timeout``.
+
+    This handles the race window after a resubscription where the server-side
+    subscription reports active but is not yet delivering messages: a message
+    published in that window is lost, so we re-publish until one arrives.
+
+    Note: this intentionally polls with the non-blocking
+    ``try_get_pubsub_message()`` instead of the blocking ``get_pubsub_message()``.
+    A blocking read would wait for a message that was never delivered (due to the
+    race above) until the client request timeout, surfacing as a TimeoutError.
+
+    Returns the received message, or ``None`` if no message arrived after
+    ``max_attempts``.
+    """
+    for attempt in range(max_attempts):
+        publishing_client.publish(message, channel)
+
+        # Poll for the message using try_get or callback check
+        poll_deadline = time.time() + poll_timeout
+        while time.time() < poll_deadline:
+            if method == MethodTesting.Callback:
+                if len(callback_messages) >= expected_idx + 1:
+                    return decode_pubsub_msg(callback_messages[expected_idx])
+            else:
+                result = listening_client.try_get_pubsub_message()
+                if result is not None:
+                    return decode_pubsub_msg(result)
+            time.sleep(poll_interval)
+
+    return None
+
 
 class TestSyncPubSub:
     @pytest.mark.parametrize("cluster_mode", [True, False])
@@ -3602,31 +3656,19 @@ class TestSyncPubSub:
             # subscription may still need a brief moment to become fully active
             # (especially in cluster mode). Use a publish-and-retry loop to
             # handle this race condition reliably.
-            msg_after = None
-            max_publish_attempts = 5
-            for attempt in range(max_publish_attempts):
-                publishing_client.publish(message_after, channel)
-
-                # Poll for the message using try_get or callback check
-                poll_deadline = time.time() + 3.0
-                while time.time() < poll_deadline:
-                    if method == MethodTesting.Callback:
-                        if len(callback_messages) >= 2:
-                            msg_after = decode_pubsub_msg(callback_messages[1])
-                            break
-                    else:
-                        result = listening_client.try_get_pubsub_message()
-                        if result is not None:
-                            msg_after = decode_pubsub_msg(result)
-                            break
-                    time.sleep(0.1)
-
-                if msg_after is not None:
-                    break
+            msg_after = _publish_and_wait_for_message(
+                publishing_client,
+                message_after,
+                channel,
+                listening_client,
+                method,
+                callback_messages,
+                expected_idx=1,
+            )
 
             assert msg_after is not None, (
                 f"Failed to receive message after reconnection "
-                f"({max_publish_attempts} publish attempts)"
+                f"({_MAX_PUBLISH_ATTEMPTS} publish attempts)"
             )
             assert msg_after.message == message_after
             assert msg_after.channel == channel
@@ -3709,31 +3751,19 @@ class TestSyncPubSub:
             # subscription may still need a brief moment to become fully active
             # (especially in cluster mode). Use a publish-and-retry loop to
             # handle this race condition reliably.
-            msg_after = None
-            max_publish_attempts = 5
-            for attempt in range(max_publish_attempts):
-                publishing_client.publish(message_after, channel)
-
-                # Poll for the message using try_get or callback check
-                poll_deadline = time.time() + 3.0
-                while time.time() < poll_deadline:
-                    if method == MethodTesting.Callback:
-                        if len(callback_messages) >= 2:
-                            msg_after = decode_pubsub_msg(callback_messages[1])
-                            break
-                    else:
-                        result = listening_client.try_get_pubsub_message()
-                        if result is not None:
-                            msg_after = decode_pubsub_msg(result)
-                            break
-                    time.sleep(0.1)
-
-                if msg_after is not None:
-                    break
+            msg_after = _publish_and_wait_for_message(
+                publishing_client,
+                message_after,
+                channel,
+                listening_client,
+                method,
+                callback_messages,
+                expected_idx=1,
+            )
 
             assert msg_after is not None, (
                 f"Failed to receive message after reconnection "
-                f"({max_publish_attempts} publish attempts)"
+                f"({_MAX_PUBLISH_ATTEMPTS} publish attempts)"
             )
             assert msg_after.message == message_after
             assert msg_after.channel == channel

